### PR TITLE
Hide Apply gift card button if gift card has not been entered

### PR DESCRIPTION
**Pull request by tflight**
We have having issues where a customer is intending to pay with a credit card, but not a gift card. They enter in their credit card details and see the 'Apply' button below and click it, thinking they need to apply their credit card number before they place the order as is a common flow on many e-commerce sites. Then they get confused when nothing happens.

This change sets the default visibility of the apply button to hidden. (It doesn't use the display property so that the button will still take up space in the DOM.) Then it applies listeners on the gift card number and pin fields. If the gift card number field is 16 digits or more and the pin field is four digits or more then the apply button is made visible.

I'm uncertain of your valid gift card number ranges so the check of `16` may not be the most appropriate. Likewise, `keyup` may or may not be the most appropriate event to hook into, though I haven't uncovered any issues yet.

I also think the button should have a label of `Apply gift card` instead of just `Apply`, but I'm doing this through a plugin that modifies the translation with the `gettext` filter. I'm also filtering the `Use a gift card` string to be `Use a Acme Company gift card` because customers were trying to use things like a Visa Gift Card in those fields.

### CODE REVIEW

File: templates/payment-fields.php [hunk 1]
```diff
@@ -143,12 +143,25 @@
...
                               window.applyGiftCard();
                           });
+                          jQuery("#gift-card-number").on('keyup', function (event) {
+                            checkEnableGiftCardApplyButton();
+                          });
+                          jQuery("#gift-card-pin").on('keyup', function (event) {
+                            checkEnableGiftCardApplyButton();
+                          });
+                          function checkEnableGiftCardApplyButton() {
+                            if (jQuery("#gift-card-number").val().length >= 16 && jQuery("#gift-card-pin").val().length >= 4) {
```
Comment (MSmedal) [inline templates/payment-fields.php:159]: The card number length should always be 19 digits and the pin length should always be 4. If we're going to validate these values then let's evaluate for those exact numbers

**Review by MSmedal — CHANGES_REQUESTED** [review]

**Comment by tflight** [comment]
I've done that now. I'm still not sure I love how this works. It doesn't give the user any feedback as they start to enter a card number or indication of what they should do next. It also doesn't give any feedback if say they've entered one digit too short on accident or one too many.

I wonder about maybe hiding the 'Apply' button by default. Then making it visible but disabled if there is any content in the gift card number field or PIN field, then finally enabling the button if the sting lengths match 19 and 4.

I still think it is an improvement over what is there now because too many of our customers were trying to "Apply" their credit card number.

**Comment by MSmedal** [comment]
> I've done that now. I'm still not sure I love how this works. It doesn't give the user any feedback as they start to enter a card number or indication of what they should do next. It also doesn't give any feedback if say they've entered one digit too short on accident or one too many.
> 
> I wonder about maybe hiding the 'Apply' button by default. Then making it visible but disabled if there is any content in the gift card number field or PIN field, then finally enabling the button if the sting lengths match 19 and 4.
> 
> I still think it is an improvement over what is there now because too many of our customers were trying to "Apply" their credit card number.

I'm unsure of what you mean. If these changes are added to the plugin then the apply button will only appear if/when both the gift card number and gift card pin have been entered completely. The apply button will also disappear if that's no longer the case. Also, the plugin displays an error message if there is an issue when the apply button is pressed or the checkout page is updated if a gift card is successfully applied to the order.

Can you clarify on whether you're seeing something different in your testing?

**Review by MSmedal — APPROVED** [review]

**Comment by tflight** [comment]
> Can you clarify on whether you're seeing something different in your testing?

I'm not seeing anything different. I'm just having more reservations about it the more I test with it. As discussed at the top e have had numerous users clicking the 'Apply' button when using only a credit card. But this may not be the right solution to that.

A more typical UI experience is that the next action button (Add to Cart, Proceed to Checkout, Place Order) is visible before validation. This just seems to leave the user hanging as to what to do next while they type in the gift card number/PIN. Yes, I realize I'm the one who suggested this change. :)

Longer term I wonder about hiding the gift card number, gift card PIN, and apply gift card button until the user clicks a checkbox indicating they want to use a gift card on the order, likely with a similar design style to the 'Pay with credit card' radio element.

I do think this PR improves the current flow, but might still need a longer look.

**Comment by MSmedal** [comment]
> I do think this PR improves the current flow, but might still need a longer look.

Agreed. We'll proceed with this change as-is in the meantime.

**Pull request merged**